source/slang/slang-ir-byte-address-legalize.cpp

```diff
@@ -205,17 +206,17 @@ struct ByteAddressBufferLegalizationContext
         return false;
     }
 
-    bool checkUnaligned(IRInst* baseOffset, IRIntegerValue immediateOffset, IRType* elementType, IRIntegerValue elementCount)
+    // Helper function to check if an IRInst is an Integer Literal
+    // TODO: Is there a better place for this generic function?
+    bool isAligned(IRInst* offset, IRIntegerValue immediateOffset, IRInst* alignment)
     {
         // Check whether the given composite resource type is aligned to the baseOffset
-        IRSizeAndAlignment elementLayout;
-        SLANG_RETURN_FALSE_ON_FAIL(getNaturalSizeAndAlignment(m_targetProgram->getOptionSet(), elementType, &elementLayout));
-        IRIntegerValue elementStride = elementLayout.getStride();
-        bool isUnaligned = true;
-        if (auto baseOffsetVal = as<IRIntLit>(baseOffset)) {
-            isUnaligned = ((baseOffsetVal->getValue() + immediateOffset) % (elementStride * elementCount)) != 0;
+        if (auto alignInst = as<IRIntLit>(alignment)) {
+            if (auto baseOffsetVal = as<IRIntLit>(offset)) {
+                return ((baseOffsetVal->getValue() + immediateOffset) % (alignInst->getValue())) == 0;
+            }
         }
-        return isUnaligned;
```

Comment:
I have a question.
Wasn't the code before your change doing the same thing that you are doing in this change?
It was checking the alignment based on the alignment of the given element type.
Is the difference on whether the alignment is based on the element type of the vector type?

Reply:
I had some misunderstanding with the expectation. now it should be fixed.